testing/base/source_gen_test.cpp

```diff
@@ -0,0 +1,171 @@
+
+// Tiny helper to sum the sizes of a range of ranges. Uses a template to avoid
+// hard coding any specific types for the two ranges.
+template <typename T>
+static auto SumSizes(const T& range) -> ssize_t {
+  ssize_t sum = 0;
+  for (const auto& inner_range : range) {
+    sum += inner_range.size();
+  }
+  return sum;
+}
+
+TEST(SourceGenTest, Ids) {
+  SourceGen gen;
+
+  auto ids = gen.GetShuffledIds(1000);
+  EXPECT_THAT(ids.size(), Eq(1000));
+  for (llvm::StringRef id : ids) {
+    EXPECT_THAT(id, MatchesRegex("[A-Za-z][A-Za-z0-9_]*"));
+  }
+
+  // We should have at least one identifier of each length [1, 64]. The exact
+  // distribution is an implementation detail designed to vaguely match the
+  // expected distribution in source code.
+  for (int size : llvm::seq_inclusive(1, 64)) {
+    EXPECT_THAT(ids, Contains(SizeIs(size)));
+  }
+
+  // Check that repeated calls are different in interesting ways, but have the
+  // exact same total bytes.
+  ssize_t ids_size_sum = SumSizes(ids);
+  for ([[maybe_unused]] int i : llvm::seq(10)) {
+    auto ids2 = gen.GetShuffledIds(1000);
+    EXPECT_THAT(ids2, SizeIs(1000));
+    // Should be (at least) a different shuffle of identifiers.
+    EXPECT_THAT(ids2, Not(ContainerEq(ids)));
+    // But the sum of lengths should be identical.
+    EXPECT_THAT(SumSizes(ids2), Eq(ids_size_sum));
+  }
+
+  // Check length constraints have the desired effect.
+  ids = gen.GetShuffledIds(1000, /*min_length=*/10, /*max_length=*/20);
+  EXPECT_THAT(ids, Each(SizeIs(AllOf(Ge(10), Le(20)))));
+
+  // Check that uniform id length results in exact coverage of each possible
+  // length for an easy case, both without and with a remainder.
+  ids = gen.GetShuffledIds(100, /*min_length=*/10, /*max_length=*/19,
+                           /*uniform=*/true);
+  EXPECT_THAT(ids, Contains(SizeIs(10)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(11)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(12)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(13)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(14)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(15)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(16)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(17)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(18)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(19)).Times(10));
+  ids = gen.GetShuffledIds(97, /*min_length=*/10, /*max_length=*/19,
+                           /*uniform=*/true);
+  EXPECT_THAT(ids, Contains(SizeIs(10)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(11)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(12)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(13)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(14)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(15)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(16)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(17)).Times(9));
+  EXPECT_THAT(ids, Contains(SizeIs(18)).Times(9));
+  EXPECT_THAT(ids, Contains(SizeIs(19)).Times(9));
```

Comment:
Are these tests for the uniform distribution in the right test function?

Reply:
I... think so?

There are two axes: unique vs. non-unique, and uniform distribution vs. non-uniform distribution.

I have a function split out for the uniqueness tests. I guess can split the two distributions into separate functions if useful?